In [74]:
import os
#read_path = 'D:/NYU-1/DL/data/chest_xray'
test_normal = 'D:/NYU-1/DL/data/chest_xray/test/NORMAL'
test_pneumonia = 'D:/NYU-1/DL/data/chest_xray/test/PNEUMONIA'
train_normal = 'D:/NYU-1/DL/data/chest_xray/train/NORMAL'
train_pneumonia = 'D:/NYU-1/DL/data/chest_xray/train/PNEUMONIA'


### train_normal

In [75]:
os.chdir(train_normal) #Change path 
normal_name_list1 = os.listdir() #Get file name 
#csv_missing=[]

In [76]:
train_n = []
for i in normal_name_list1:
    if 'NORMAL2' in i:
        i = i[8:]
    patient = i[:7]
    train_n.append(patient)
train_normal_distinct = list(set(train_n))
print(len(train_n))
print(len(train_normal_distinct))

1349
1008


### test_normal

In [78]:
os.chdir(test_normal) #Change path 
normal_name_list2 = os.listdir() #Get file name 
#csv_missing=[]

In [79]:
test_n = []
for i in normal_name_list2:
    if 'NORMAL2' in i:
        i = i[8:]
    patient = i[:7]
    test_n.append(patient)
test_normal_distinct = list(set(test_n))
print(len(test_n))
print(len(test_normal_distinct))

234
202


In [80]:
#patient id in normal dataset, eg. IM-0001
normal_all = list(set(train_normal_distinct+test_normal_distinct))
print(len(normal_all))

1116


In [81]:
#check
count = 0
for i in train_normal_distinct:
    if i in test_normal_distinct:
        count+=1
print(count)

#1008+202-94=1116

94


In [83]:

for i in normal_name_list1:
    if i in normal_name_list2:
        print(i)

### train_pneumonia

In [ ]:
os.chdir(train_pneumonia) #Change path 
normal_name_list3 = os.listdir() #Get file name 
#csv_missing=[]

In [84]:
train_p = []
for i in normal_name_list3:
    patient = i.split('_')[0]
    train_p.append(patient)
    #train_normal.append(patient)
train_pneumonia_distinct = list(set(train_p))
print(len(train_p))
print(len(train_pneumonia_distinct))

3883
1642


### test_pneumonia

In [85]:
os.chdir(test_pneumonia) #Change path 
normal_name_list4 = os.listdir() #Get file name 
#csv_missing=[]

In [86]:
test_p = []
for i in normal_name_list4:
    patient = i.split('_')[0]
    test_p.append(patient)
    #train_normal.append(patient)
test_pneumonia_distinct = list(set(test_p))
print(len(test_p))
print(len(test_pneumonia_distinct))

390
202


In [87]:
for i in normal_name_list3:
    if i in normal_name_list4:
        print(i)

In [146]:
pneumonia_all = list(set(train_pneumonia_distinct+test_pneumonia_distinct))
print(len(pneumonia_all))

1674


In [93]:
# check
count = 0
for i in train_pneumonia_distinct:
    if i in test_pneumonia_distinct:
        count+=1
print(count)

#1642+202-170=1674

170


### normal split

In [113]:
print(normal_all[:5])

['IM-0948', 'IM-1046', 'IM-0898', 'IM-1039', 'IM-0732']


In [114]:
# all normal image name
normal_list = normal_name_list1 + normal_name_list2
import random
random.shuffle(normal_all)
print(normal_all[:5])

['IM-0826', 'IM-0440', 'IM-0019', 'IM-0239', 'IM-0828']


In [115]:
train_n_id = normal_all[:781] #70%
test_n_id = normal_all[781:893] #10%
val_n_id = normal_all[893:] #20%

In [126]:
train_n_list = [] #image name

for i in train_n_id:
    for j in normal_list:
        if i in j:
            train_n_list.append(j)
            
print(len(train_n_list))

1095


In [127]:
test_n_list = [] #image name

for i in test_n_id:
    for j in normal_list:
        if i in j:
            test_n_list.append(j)
    
print(len(test_n_list))

159


In [128]:
val_n_list = [] #image name

for i in val_n_id:
    for j in normal_list:
        if i in j:
            val_n_list.append(j)
            
print(len(val_n_list))

329


In [ ]:
#1095+159+329 = 1583, equal to normal image numbers

### pneumonia split

In [147]:
print(pneumonia_all[:5])
pneumonia_list = normal_name_list3 + normal_name_list4
random.shuffle(pneumonia_all)
print(pneumonia_all[:5])

['person634', 'person1764', 'person1746', 'person632', 'person448']
['person299', 'person1408', 'person1847', 'person802', 'person17']


In [149]:
for i in range(len(pneumonia_all)):
    id_ = pneumonia_all[i]+'_'
    pneumonia_all[i] = id_
print(pneumonia_all[5])
    
    

person531_


In [150]:
train_p_id = pneumonia_all[:1172] #70%
test_p_id = pneumonia_all[1172:1339] #10%
val_p_id = pneumonia_all[1339:] #20%

In [160]:
train_p_list = [] #image name

for i in train_p_id:
    for j in pneumonia_list:
        if i in j:
            train_p_list.append(j)
            
print(len(train_p_list))

2949


In [161]:
test_p_list = [] #image name

for i in test_p_id:
    for j in pneumonia_list:
        if i in j:
            test_p_list.append(j)
            
print(len(test_p_list))

443


In [162]:
val_p_list = [] #image name

for i in val_p_id:
    for j in pneumonia_list:
        if i in j:
            val_p_list.append(j)
            
print(len(val_p_list))

881


### new table: indicating train/val/test 

In [174]:
import pandas as pd
data1 = {'image_index':train_p_list,
         'class':'pneumonia',
         'label': 1}
data2 = {'image_index':train_n_list,
         'class':'normal',
         'label': 0}
df1 = pd.DataFrame(data1)
df2 = pd.DataFrame(data2)
df_train = pd.concat([df1, df2], axis = 0)
print(df_train.shape)
df_train = df_train.sample(frac = 1)
df_train.head()


(4044, 3)


image_index      class  label
0  person299_bacteria_1414.jpeg  pneumonia      1
1  person299_bacteria_1416.jpeg  pneumonia      1
2  person299_bacteria_1417.jpeg  pneumonia      1
3  person299_bacteria_1418.jpeg  pneumonia      1
4  person299_bacteria_1419.jpeg  pneumonia      1

In [175]:
data3 = {'image_index':test_p_list,
         'class':'pneumonia',
         'label': 1}
data4 = {'image_index':test_n_list,
         'class':'normal',
         'label': 0}
df3 = pd.DataFrame(data3)
df4 = pd.DataFrame(data4)
df_test = pd.concat([df3, df4], axis = 0)
df_test = df_test.sample(frac = 1)
df_test.shape

(602, 3)

In [177]:
data5 = {'image_index':val_p_list,
         'class':'pneumonia',
         'label': 1}
data6 = {'image_index':val_n_list,
         'class':'normal',
         'label': 0}
df5 = pd.DataFrame(data5)
df6 = pd.DataFrame(data6)
df_val = pd.concat([df5, df6], axis = 0)
df_val = df_val.sample(frac = 1)
df_val.shape

(1210, 3)

In [179]:
df_train.to_csv('train.csv', index = False)
df_val.to_csv('val.csv', index = False)
df_test.to_csv('test.csv', index = False)

**train: (70% patients)**
normal: 1095 images, 781 patients

pneumonia: 2949 images, 1172 patients

normal image/train image: 0.271

**test: (10% patients)**

normal: 159 images, 112 patients

pneumonia: 443 images, 167 patients

normal image/test image: 0.264

**validation: (20% patients)**

normal: 329 images, 223 patients

pneumonia: 881 images, 335 patients

normal image/val image: 0.272

NO DATA LEAKAGE!